SAMtags.tex

```diff
@@ -453,6 +455,112 @@ \subsubsection{Color space}
+Xanthosine (n), irrespective of the sequence composition.
+
+The standard code types and their associated ChEBI values are listed
+below, taken from \emph{Modeling methyl-sensitive transcription factor
+motifs with an expanded epigenetic alphabet}, Coby Viner
+et.al. \url{https://www.biorxiv.org/content/10.1101/043794v1}.
+Additionally ambiguity codes {\tt A}, {\tt C}, {\tt G} and {\tt T}
+exist to represent unspecified modifications bases of their respective
+canonical base types, plus code {\tt N} to represent an unspecified
+modification of any base type.
+
+\begin{center}
+\begin{tabular}{lllll}
+{\bf Unmodified base} & {\bf Code} & {\bf Abbreviation} & {\bf Name} & {\bf ChEBI} \\
+\hline
+C & m & 5mC   & 5-Methylcytosine        & 27551 \\
+C & h & 5hmC  & 5-Hydroxymethylcytosine & 76792 \\
+C & f & 5fC   & 5-Formylcytosine        & 76794 \\
+C & c & 5caC  & 5-Carboxylcytosine      & 76793 \\
+C & C &       & Ambiguity code; any C mod & \\
+\hline
+T & g & 5hmU  & 5-Hydroxymethyluracil   & 16964 \\
+T & e & 5fU   & 5-Formyluracil          & 80961 \\
+T & b & 5caU  & 5-Carboxyluracil        & 17477 \\
+T & T &       & Ambiguity code; any T mod & \\
+\hline
+A & a & 6mA   & 6-Methyladenine         & 28871 \\
+A & A &       & Ambiguity code; any A mod & \\
+\hline
+G & o & 8oxoG & 8-Oxoguanine            & 44605 \\
+G & G &       & Ambiguity code; any G mod & \\
+\hline
+N & n & Xao   & Xanthosine              & 18107 \\
+N & N &       & Ambiguity code; any mod & \\
+\end{tabular}
+\end{center}
+
+\item[MP:Z:\tagvalue{qualities}]
+\hfill\\
+The optional {\tt MP} tag lists the Phred qualities of each modification listed in the {\tt MM} tag.
+The qualities are encoded in the same manner as the primary {\sf QUAL} field; one byte per quality with ASCII value Phred score + 33.
+A space character (`{\tt \textvisiblespace}') should be used as a separator between concatenated quality strings when multiple modification lists are present in the {\tt MM} tag.
+The length should match the number of position deltas from {\tt MM} plus 1 per space character required.
+
+For example {\tt MM:Z:C+m,5,12,3;C+h,57;} may have an associated
+quality tag of {\tt MP:Z:5EB /}.
+
+Quality values for ambiguity codes give the likelihood that the
+modification is one of the possible codes compatible with that
+ambiguity code.  For example {\tt MM:Z:C+C,10; MP:Z:+} indicates a C
+call with an unspecified modification and the phred score of 10 (ASCII
+value {\tt +}).  This corresponds to a 90\% chance of the base being
+modified.
+
+To represent several possible modifications at the same site the {\tt MP} tag can be used to indicate the probabilities of each possibility.
+The values used should be absolute probabilities, not relative between the alternatives.
+For example, a C base that has 90\% chance of being modified with 5mC being three times more likely than 5hmC will encode 5mC with 67.5\% probability ($0.9 * 0.75$)and 5hmC with 22.5\% probability ($0.9 * 0.25$).
```

Comment:
If we really want to be able to encode probabilities in the range of 10% - 90% perhaps we should not use Phred which has terrible resolution in that range (10%=10 and 90%=".45"...1?0?)
Please be explicit that the `"` and `&` are encoding phred 1 and 5.

Reply:
At the time of writing this I hadn't seen any real data.

However I thought about that once I saw how ONT handle things.  They report probabilities as p*255; ie p=0.33 is 84.  They have arrays of values, so if trained on C, 5mC and 5hmC then they'd have 3 score for C which jointly add up to 255 (irrespective of whether they believe the call to be C - this is to be used after combining with the base call probabilities themselves), giving the relative likelihoods of the various modified and unmodified forms.

However it also means they cannot give higher certainties than 99.6%, or around phred 24.  That may well be sufficient.

The question is, do we acknowledge this is different data and is never going to be extremely accurate?   Do we actually care about small fluctuations in probabilities at the low end?  Eg p=0.15 vs p=0.2?

Reply:
Actually it's simply not possible, which comes down to the phred range again.

If we wish to emit probabilities for more than one possibility, which I believe is how ONT Guppy may go in the future (https://twitter.com/AlbertVilella/status/1151395294879395845), then phred scores will always generate a score of 0, 1 or 2 somewhere.  That's because inherently there will be one call with p < 0.5.

The Illumina answer to this was to do phred+64 and use log-odds instead of log.  Ie log(p/(1-p)) instead of log(1-p).  That gives 0 being p=0.5 and is symmetric so choice A if p=.99 and choice B of p=0.01 gives phred scores of +20 and -20.

However it still begs the question of whether a log scale is appropriate, or whether even knowing the other values helps.  When I was doing consensus calling algorithms I always grumbled that phred isn't good enough.  If you're trying to make a call and you get A with P=0.9 but everything else is C, then it really helps to know whether than remaining 0.1 probabiity on the A was evenly spread across C, G and T or predominantly a remainder for C.  (I just had to go with flat distribution in lieu of anything better.)

I simply don't have enough understanding of the downstream tools to know whether emitting the score for the most likely call is correct (the phred approach), or whether we the benefits of knowing 2nd, 3rd best choices outweighs the extra storage cost.